### PR TITLE
Created a simple DAG RNN which the more complex DAG RNN now extends

### PR DESCRIPTION
**Pull request by erikhuck**

### CODE REVIEW

File: dna/models.py [hunk 1]
```diff
@@ -413,21 +404,18 @@ def __init__(
         self.lstm.to(device=self.device)
 
         lstm_output_size = hidden_state_size * n_directions
+        output_submodule_input_size = lstm_output_size + metafeatures_length
+
+        if isinstance(self, DAGRNN):
```
Comment (bjschoenfeld) [inline dna/models.py]: This is probably not good design. A parent class has no knowledge of its children classes.

File: dna/models.py [hunk 2]
```diff
@@ -445,20 +433,18 @@ def forward(self, args):
...
-        hidden_state = torch.cat(hidden_state, dim=0)
+        hidden_state = torch.zeros(self.hidden_state_dim0_size, batch_size, self.hidden_state_size, device=self.device)
         cell_state = hidden_state
         lstm_start_state = (hidden_state, cell_state)
 
+
+        dag_rnn_output = self.parse_DAG(seq_len, pipelines, pipeline_structure, lstm_start_state)
+        linear_input = torch.cat((dag_rnn_output, metafeatures), dim=1)
+
+        linear_output = self._output_submodule(linear_input)
```
Comment (bjschoenfeld) [inline dna/models.py]: Why are these variables called "linear"? This are hidden states of the network. Also, the output submodule is probably not linear either. Perhaps there is a better name.

File: model_configs/dagrnn_regression_config.json [hunk 3]
```diff
@@ -1,5 +1,6 @@
 {
     "__init__": {
+        "basic": false,
```
Comment (bjschoenfeld) [inline model_configs/dagrnn_regression_config.json]: The BasicDAGRNN and DAGRNN are different models. Perhaps they should not be controlled with a single bool in a config file. Rather, let's make them both first class citizens and add the basic one to the test.sh script, give it its own model class, etc.

**Comment by bjschoenfeld** [comment]
Since there are two different models that use a DAGRNN over the pipeline, it might be a good idea to make  a separate torch.nn.module class for the DAGRNN that can be used by any model. It could have a method for getting the initial hidden state, which could be subclassed for when the metafeatures are the hidden state.

Also, the model you are now calling "BasicDAGRNN" could just be called "DAGRNN" and the model that uses the metafeatures for the initial hidden state should be called something that suggests this special use. Thoughts?

File: dna/models.py [hunk 4]
```diff
@@ -628,16 +671,13 @@ def encode_pipeline(self, pipeline, primitive_to_enc):
         return encoding
 
     def _get_model(self, train_data):
+        print('Using Basic DAG RNN')
```
Comment (bjschoenfeld) [inline dna/models.py]: let's keep the print statements to a minimum.

File: dna/models.py [hunk 5]
```diff
@@ -663,6 +703,35 @@ def _get_loss_function(self):
...
+        super().__init__(activation_name, hidden_state_size, lstm_n_layers, lstm_dropout, bidirectional,
+                         output_n_hidden_layers, output_hidden_layer_size, output_dropout, use_batch_norm,
+                         use_skip, device=device, seed=seed)
+
+        self.input_n_hidden_layers = input_n_hidden_layers
+        self.input_hidden_layer_size = input_hidden_layer_size
+        self.input_dropout = input_dropout
+
+    def _get_model(self, train_data):
+        print('Using Complex DAG RNN')
```
Comment (bjschoenfeld) [inline dna/models.py]: let's remove this

File: dna/models.py [hunk 6]
```diff
@@ -475,27 +455,18 @@ def forward(self, args):
...
 
-        # Initialize the hidden state and cell state using a concatenation of the metafeatures
-        hidden_state = []
-        for i in range(self.hidden_state_dim0_size):
-            hidden_state.append(metafeatures)
-        hidden_state = torch.cat(hidden_state, dim=0)
-        cell_state = hidden_state
-        lstm_start_state = (hidden_state, cell_state)
+        fc_output = self._output_submodule(fc_input)
+        return fc_output.squeeze()
```
Comment (bjschoenfeld) [inline dna/models.py]: Let's talk about the output of the forward function of the DAGRNN.

**Comment by bjschoenfeld** [comment]
@erikhuck Can we merge develop into this branch together?

**Comment by erikhuck** [comment]
> @erikhuck Can we merge develop into this branch together?

@bjschoenfeld yes

**Comment by bjschoenfeld** [comment]
It seems that the `bidirectional` feature of the DAGLSTM is not truly being used. We are stepping through the DAG one node at a time, only in the forward direction. The LSTM is only seeing one node at time, so when `bidirectional` is set, the LSTM will output the same outputs twice.

Bidirectionality is not as meaningful in our context either, so I am going to remove the it.

**Comment by erikhuck** [comment]
I remember bidirectional empirically producing better results.

File: dna/models.py [hunk 7]
```diff
@@ -423,67 +423,36 @@ def __init__(
...
-                lstm_input_state = self.get_lstm_input_state(
-                    prev_lstm_states=prev_lstm_states, primitive_inputs=primitive_inputs
-                )
-
-            (lstm_output, lstm_output_state) = self.lstm(encoded_primitives, lstm_input_state)
-
-            prev_lstm_states.append(lstm_output_state)
+        prev_lstm_states = {'inputs.0': lstm_start_state}  # TODO: use something better than 'inputs.0'
+        for i, node_inputs in enumerate(dag_structure):
+            # TODO: use LSTMCell
```
Comment (erikhuck) [inline dna/models.py]: What is "LSTMCELL"?

**Comment by bjschoenfeld** [comment]
What is it about the bidirectionality that made it better? Is it just more weights?

File: dna/models.py (hunk 7 above)
Comment (bjschoenfeld) [reply dna/models.py]: https://pytorch.org/docs/stable/nn.html#lstmcell

It is the core component of an LSTM. It is the component that is used at each step in a sequence. It takes as input the hidden state, the cell state, and input feature vector at time step t and outputs  new hidden and cell states as well as the output feature vector.

Comment (bjschoenfeld) [reply dna/models.py]: It is essentially an LSTM for a sequence of length 1.

**Comment by bjschoenfeld** [comment]
@erikhuck Can you review this now?

File: dna/models.py (hunk 7 above)
Comment (erikhuck) [reply dna/models.py]: Cool

**Pull request merged**